### PR TITLE
pythonPackages.lookatme: reintroduce at 2.5.0

### LINKED ISSUE #208969: Packaging request: lookatme
**Project description**

"An interactive, terminal-based markdown presenter"

**Metadata**

* homepage URL: https://lookatme.readthedocs.io/en/latest/
* source URL: https://github.com/d0c-s4vage/lookatme
* license: mit
* platforms: unix (python)

### PR DESCRIPTION
**Pull request by totoroot**
###### Description of changes

Add [lookatme](https://github.com/d0c-s4vage/lookatme), an interactive, extensible, terminal-based markdown presentation tool.

It got removed earlier this year in PR #170162 due to a Python2 dependency with click version 7.
The [dependencies got relaxed](https://github.com/d0c-s4vage/lookatme/pull/141), so click version 8 is now allowed as well.

I added the former maintainer @AmeerTaweel to lookatme's maintainers as well as myself as I've been using it now and then recently.
I also added tests. It builds fine on x86_64-linux, and I'll test building on x86_64-darwin soon as well.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

closes https://github.com/NixOS/nixpkgs/issues/208969

### CODE REVIEW

**Comment by totoroot** [comment]
Result of `nixpkgs-review pr 209302` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>python310Packages.hug</li>
    <li>python39Packages.hug</li>
  </ul>
</details>
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>python39Packages.lookatme</li>
  </ul>
</details>
<details>
  <summary>72 packages built:</summary>
  <ul>
    <li>ansible-doctor (python310Packages.ansible-doctor)</li>
    <li>apache-airflow (python310Packages.apache-airflow)</li>
    <li>esphome</li>
    <li>faraday-agent-dispatcher</li>
    <li>flashfocus</li>
    <li>hassil (python310Packages.hassil)</li>
    <li>home-assistant-component-tests.renault</li>
    <li>home-assistant-component-tests.sia</li>
    <li>json-schema-for-humans (python310Packages.json-schema-for-humans)</li>
    <li>localstack (python310Packages.localstack)</li>
    <li>lookatme (python310Packages.lookatme)</li>
    <li>platformio</li>
    <li>python310Packages.aiohttp-apispec</li>
    <li>python310Packages.aiomysensors</li>
    <li>python310Packages.apispec</li>
    <li>python310Packages.cairo-lang</li>
    <li>python310Packages.dataclasses-json</li>
    <li>python310Packages.deal</li>
    <li>python310Packages.deezer-python</li>
    <li>python310Packages.environs</li>
    <li>python310Packages.faraday-agent-parameters-types</li>
    <li>python310Packages.flask-appbuilder</li>
    <li>python310Packages.flask_marshmallow</li>
    <li>python310Packages.gradient</li>
    <li>python310Packages.icontract</li>
    <li>python310Packages.lektor</li>
    <li>python310Packages.marshmallow</li>
    <li>python310Packages.marshmallow-dataclass</li>
    <li>python310Packages.marshmallow-enum</li>
    <li>python310Packages.marshmallow-oneofschema</li>
    <li>python310Packages.marshmallow-polyfield</li>
    <li>python310Packages.marshmallow-sqlalchemy</li>
    <li>python310Packages.pylddwrap</li>
    <li>python310Packages.pysiaalarm</li>
    <li>python310Packages.pytenable</li>
    <li>python310Packages.renault-api</li>
    <li>python310Packages.spyse-python</li>
    <li>python310Packages.vaa</li>
    <li>python310Packages.webargs</li>
    <li>python39Packages.aiohttp-apispec</li>
    <li>python39Packages.aiomysensors</li>
    <li>python39Packages.ansible-doctor</li>
    <li>python39Packages.apache-airflow</li>
    <li>python39Packages.apispec</li>
    <li>python39Packages.cairo-lang</li>
    <li>python39Packages.dataclasses-json</li>
    <li>python39Packages.deal</li>
    <li>python39Packages.deezer-python</li>
    <li>python39Packages.environs</li>
    <li>python39Packages.faraday-agent-parameters-types</li>
    <li>python39Packages.flask-appbuilder</li>
    <li>python39Packages.flask_marshmallow</li>
    <li>python39Packages.gradient</li>
    <li>python39Packages.hassil</li>
    <li>python39Packages.icontract</li>
    <li>python39Packages.json-schema-for-humans</li>
    <li>python39Packages.lektor</li>
    <li>python39Packages.localstack</li>
    <li>python39Packages.marshmallow</li>
    <li>python39Packages.marshmallow-dataclass</li>
    <li>python39Packages.marshmallow-enum</li>
    <li>python39Packages.marshmallow-oneofschema</li>
    <li>python39Packages.marshmallow-polyfield</li>
    <li>python39Packages.marshmallow-sqlalchemy</li>
    <li>python39Packages.pylddwrap</li>
    <li>python39Packages.pysiaalarm</li>
    <li>python39Packages.pytenable</li>
    <li>python39Packages.renault-api</li>
    <li>python39Packages.vaa</li>
    <li>python39Packages.webargs</li>
    <li>sublime-music</li>
    <li>tribler</li>
  </ul>
</details>

**Comment by totoroot** [comment]
`lookatme` fails to build for Python versions < 3.10 with the following error and I have no idea how to get rid of it.
I've looked for the solution for too long then I'd like to admit, but I cannot understand why the Python 3.8 and 3.9 environments do not include setuptools :exploding_head: 

    error: builder for '/nix/store/lj1fn76fipyklpiwqwynqpmdp2kf35nd-python3.9-lookatme-2.5.0.drv' failed with exit code 1;
           last 10 log lines:
           > setting SOURCE_DATE_EPOCH to timestamp 315619200 of file source/tox.ini
           > patching sources
           > configuring
           > no configure script, doing nothing
           > building
           > Executing setuptoolsBuildPhase
           > Traceback (most recent call last):
           >   File "/build/source/nix_run_setup", line 3, in <module>
           >     import setuptools
           > ModuleNotFoundError: No module named 'setuptools'
           For full logs, run 'nix log /nix/store/lj1fn76fipyklpiwqwynqpmdp2kf35nd-python3.9-lookatme-2.5.0.drv'.

**Comment by totoroot** [comment]
Result of `nixpkgs-review pr 209302` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages marked as broken and skipped:</summary>
  <ul>
    <li>python310Packages.cairo-lang</li>
    <li>python310Packages.hug</li>
    <li>python39Packages.cairo-lang</li>
    <li>python39Packages.hug</li>
  </ul>
</details>
<details>
  <summary>30 packages failed to build:</summary>
  <ul>
    <li>python39Packages.aiohttp-apispec</li>
    <li>python39Packages.ansible-doctor</li>
    <li>python39Packages.apache-airflow</li>
    <li>python39Packages.apispec</li>
    <li>python39Packages.dataclasses-json</li>
    <li>python39Packages.deal</li>
    <li>python39Packages.deezer-python</li>
    <li>python39Packages.environs</li>
    <li>python39Packages.faraday-agent-parameters-types</li>
    <li>python39Packages.flask-appbuilder</li>
    <li>python39Packages.flask_marshmallow</li>
    <li>python39Packages.gradient</li>
    <li>python39Packages.hassil</li>
    <li>python39Packages.icontract</li>
    <li>python39Packages.json-schema-for-humans</li>
    <li>python39Packages.lektor</li>
    <li>python39Packages.localstack</li>
    <li>python39Packages.lookatme</li>
    <li>python39Packages.marshmallow</li>
    <li>python39Packages.marshmallow-dataclass</li>
    <li>python39Packages.marshmallow-enum</li>
    <li>python39Packages.marshmallow-oneofschema</li>
    <li>python39Packages.marshmallow-polyfield</li>
    <li>python39Packages.marshmallow-sqlalchemy</li>
    <li>python39Packages.pylddwrap</li>
    <li>python39Packages.pysiaalarm</li>
    <li>python39Packages.pytenable</li>
    <li>python39Packages.renault-api</li>
    <li>python39Packages.vaa</li>
    <li>python39Packages.webargs</li>
  </ul>
</details>
<details>
  <summary>32 packages built:</summary>
  <ul>
    <li>ansible-doctor (python310Packages.ansible-doctor)</li>
    <li>apache-airflow (python310Packages.apache-airflow)</li>
    <li>faraday-agent-dispatcher</li>
    <li>hassil (python310Packages.hassil)</li>
    <li>json-schema-for-humans (python310Packages.json-schema-for-humans)</li>
    <li>localstack (python310Packages.localstack)</li>
    <li>lookatme (python310Packages.lookatme)</li>
    <li>python310Packages.aiohttp-apispec</li>
    <li>python310Packages.apispec</li>
    <li>python310Packages.dataclasses-json</li>
    <li>python310Packages.deal</li>
    <li>python310Packages.deezer-python</li>
    <li>python310Packages.environs</li>
    <li>python310Packages.faraday-agent-parameters-types</li>
    <li>python310Packages.flask-appbuilder</li>
    <li>python310Packages.flask_marshmallow</li>
    <li>python310Packages.gradient</li>
    <li>python310Packages.icontract</li>
    <li>python310Packages.lektor</li>
    <li>python310Packages.marshmallow</li>
    <li>python310Packages.marshmallow-dataclass</li>
    <li>python310Packages.marshmallow-enum</li>
    <li>python310Packages.marshmallow-oneofschema</li>
    <li>python310Packages.marshmallow-polyfield</li>
    <li>python310Packages.marshmallow-sqlalchemy</li>
    <li>python310Packages.pylddwrap</li>
    <li>python310Packages.pysiaalarm</li>
    <li>python310Packages.pytenable</li>
    <li>python310Packages.renault-api</li>
    <li>python310Packages.spyse-python</li>
    <li>python310Packages.vaa</li>
    <li>python310Packages.webargs</li>
  </ul>
</details>

File: pkgs/development/python-modules/marshmallow/default.nix [hunk 1]
```diff
@@ -10,7 +10,7 @@
 
 buildPythonPackage rec {
   pname = "marshmallow";
-  version = "3.16.0";
```
Comment (AmeerTaweel) [inline pkgs/development/python-modules/marshmallow/default.nix:13]: Is it okay to just update this? It might break other packages...

Comment (totoroot) [reply pkgs/development/python-modules/marshmallow/default.nix]: As you can see in my nixpkgs-review comments, it indeed breaks a bunch of packages on x86_64-darwin, so I dropped the commit to update marshmallow and just added the dependency itself in a compatible version.

**Comment by totoroot** [comment]
@AmeerTaweel Any ideas why building would fail for Python version < 3.10 ?

See https://github.com/NixOS/nixpkgs/pull/209302#issuecomment-1374027225

**Comment by AmeerTaweel** [comment]
@totoroot I think this is an issue with `nixpkgs`, not `lookatme`.

[Python packages are not building “ModuleNotFoundError: No module named ‘setuptools’”](https://discourse.nixos.org/t/python-packages-are-not-building-modulenotfounderror-no-module-named-setuptools/22232)

[nixpkgs#poetry is broken at nixpkgs-unstable](https://github.com/NixOS/nixpkgs/issues/193657)

**Review by ameertaweel — APPROVED** [review]

File: pkgs/development/python-modules/lookatme/default.nix [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+, pytest-mock
+, pytestCheckHook
+, six
+}:
+
+let
+  python3_ = python3.override {
+    packageOverrides = self: super: {
+      # Requires marshmallow<4,>=3.17.0
+      # Currently marshmallow is version 3.16.0 in nixpkgs
```
Comment (dotlambda) [inline pkgs/development/python-modules/lookatme/default.nix:19]: We now have 3.19.0

File: pkgs/development/python-modules/lookatme/default.nix [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        src = fetchFromGitHub {
+          owner = "marshmallow-code";
+          repo = "marshmallow";
+          rev = "refs/tags/${version}";
+          sha256 = "sha256-b1brLHM48t45bwUXk7QreLLmvTzU0sX7Uoc1ZAgGkrE=";
+        };
+      });
+      # Requires mistune>=0.8,<1
+      mistune = super.mistune.overridePythonAttrs (oldAttrs: rec {
+        name = "${oldAttrs.pname}-${version}";
```
Comment (dotlambda) [inline pkgs/development/python-modules/lookatme/default.nix:33]:
```suggestion
```

File: pkgs/development/python-modules/lookatme/default.nix [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+        version = "0.8.4";
+        src = fetchFromGitHub {
+          owner = "lepture";
+          repo = "mistune";
+          rev = "refs/tags/v${version}";
+          sha256 = "sha256-H9L2cJZVWvcbcWAF8ZMLGJGE7DO1h2wNZCbeFA02p7g=";
```
Comment (dotlambda) [inline pkgs/development/python-modules/lookatme/default.nix:39]:
```suggestion
          hash = "sha256-H9L2cJZVWvcbcWAF8ZMLGJGE7DO1h2wNZCbeFA02p7g=";
```

File: pkgs/development/python-modules/lookatme/default.nix [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+        };
+      });
+      # Requires PyYAML>=5,<6
+      pyyaml = super.pyyaml.overridePythonAttrs (oldAttrs: rec {
+        name = "${oldAttrs.pname}-${version}";
```
Comment (dotlambda) [inline pkgs/development/python-modules/lookatme/default.nix:44]:
```suggestion
```

File: pkgs/development/python-modules/lookatme/default.nix [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        version = "5.4.1.1";
+        src = fetchFromGitHub {
+          owner = "yaml";
+          repo = "pyyaml";
+          rev = "refs/tags/${version}";
+          sha256 = "sha256-qLdAMqoyEXRIqcNuHBBtST8GWh5gmx5fBU/q3f4zaOw=";
```
Comment (dotlambda) [inline pkgs/development/python-modules/lookatme/default.nix:50]:
```suggestion
          hash = "sha256-qLdAMqoyEXRIqcNuHBBtST8GWh5gmx5fBU/q3f4zaOw=";
```

File: pkgs/development/python-modules/lookatme/default.nix [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+, click
+, pygments
+, urwid
+, pytest-mock
+, pytestCheckHook
+, six
+}:
+
+let
+  python3_ = python3.override {
```
Comment (dotlambda) [inline pkgs/development/python-modules/lookatme/default.nix:16]: If you override `python3` then you cannot put it in `python-packages.nix`.

File: pkgs/top-level/all-packages.nix [hunk 8]
```diff
@@ -8790,6 +8790,8 @@ with pkgs;
 
   lolcat = callPackage ../tools/misc/lolcat { };
 
+  lookatme = python3Packages.callPackage ../development/python-modules/lookatme { };
```
Comment (dotlambda) [inline pkgs/top-level/all-packages.nix:8793]:
```suggestion
  lookatme = callPackage ../development/python-modules/lookatme { };
```

**Review by dotlambda — CHANGES_REQUESTED** [review]

**Comment by joriscode** [comment]
@totoroot Do you need help regarding testing this on Mac?

**Comment by NickCao** [comment]
Upstream project no longer actively maintained.

**Pull request closed without merging**